toolchain for ESP32 added

gschorcht opened the pull request:
I have added the installation of the ESP32 toolchain as required for the ESP32 port of RIOT,  PR https://github.com/RIOT-OS/RIOT/pull/9426.

The toolchain is installed in ```/opt/esp``` and requires about 181 MB after cleanup. All path variables in the makefiles of the ESP32 port point to this directory.

gschorcht commented:
Can we proceed with this PR to get murdock working for ESP32? Furthermore, I would like to reference ```riotdocker``` as possible toolchain option in the board documentation I'm writing for the ESP32 port.

kaspar030 commented:
looks ok. just one thing: why is python-serial and the dialout group needed? The container up to now has only be used for building. Flashing and terminal should be executed on the host.

gschorcht commented:
@kaspar030 
Ok I see, the intention of ```riotdocker``` was only to build the ELF file and not flash it.

Hm, before RIOT can be flashed, the ELF file has to be converted into a special binary image format using ```esptool.py``` which requires ```python-serial```. Flashing is also done with ```esptool.py```. Not having the flashing capability in ```riotdocker``` would require that the user has to install ```esptool.py``` on the host system.

Since older Linux distributions, e.g. Ubuntu 16.4, not offering it as a package, the user would have to start to install toolchain components. Sure, it's a documentation problem, but it does not seem to be comfortable.

kaspar030 commented:
> Hm, before RIOT can be flashed, the ELF file has to be converted into a special binary image format using esptool.py which requires python-serial. Flashing is also done with esptool.py. Not having the flashing capability in riotdocker would require that the user has to install esptool.py on the host system.

Ok, having the tools in the container won't hurt.

But the dialout group thing I don't quite get - the gid of the dialout group will certainly be container-specific, it won't (necessarily) match the dialout group of the host system. If the intention was that a dialout-owned host serial device can be used by the riotbuild user within the container, it will probably not work. (maybe with ubuntu host, ubuntu docker base image...)

miri64 commented:
I think what usually happens with other devices that the user is required to configure this properly on their host system (e.g. using `udev` and/or by joining whatever group the system assigns to the device).

gschorcht commented:
@kaspar030 What do we do next? Should I remove the addition of the riotbuild user to the dialout group?

At least for my ESP ports, I've documented that using riotdocker consists of two steps

1. making the RIOT application in Docker with command ```make BOARD= ...```
2. flashing the RIOT application on the host computer with command ```make flash-only BOARD=...```

where step 2 requires that the flasher program is installed on the host and how ```esptool.py``` has to be installed.

gschorcht commented:
@kaspar030 I removed adding the user to the group dial.

I also had to change the download procedure for the Xtensa ESP32 Toolchain. I replaced the original Espressif toolchain with a version that I created myself using crosstool-NG. This was necessary because Espressif's original toolchain uses a newlib version that printf for floats does not support.

I know Riotdocker is currently blocked by issue #46. I also went into this problem. I hope PR #47 can be merged soon and we can continue with this PR to get murdock working for ESP32.

cladmi commented on Dockerfile:
```diff
@@ -138,6 +139,32 @@ RUN echo 'Adding esp8266 toolchain' >&2 && \
+RUN echo 'Installing ESP32 toolchain' >&2 && \
+    mkdir -p /opt/esp && \
+    cd /opt/esp && \
+    git clone --recursive https://github.com/espressif/esp-idf.git && \
```
Is there a reason to not update to https://github.com/espressif/esp-idf/releases/tag/v3.1 now that it is released ?

cladmi replied:
Thanks for the precision, so yeah there is a reason to stay with is one for the moment.

leandrolanzieri commented:
The MIPS toolchain is now working. Please rebase

gschorcht commented:
> The MIPS toolchain is now working. Please rebase

@leandrolanzieri Done

jia200x commented:
I'm testing it against RIOT-OS/RIOT#9426

jia200x commented:
Toolchain works:
```
Building application "hello-world" for "esp32-olimex-evb" with MCU "esp32".

"make" -C /data/riotbuild/riotbase/core
"make" -C /data/riotbuild/riotbase/drivers
"make" -C /data/riotbuild/riotbase/drivers/mtd
"make" -C /data/riotbuild/riotbase/drivers/periph_common
"make" -C /data/riotbuild/riotbase/sys
"make" -C /data/riotbuild/riotbase/sys/auto_init
"make" -C /data/riotbuild/riotbase/sys/div
"make" -C /data/riotbuild/riotbase/sys/isrpipe
"make" -C /data/riotbuild/riotbase/sys/log
"make" -C /data/riotbuild/riotbase/sys/luid
"make" -C /data/riotbuild/riotbase/sys/newlib_syscalls_default
"make" -C /data/riotbuild/riotbase/sys/posix/pthread
"make" -C /data/riotbuild/riotbase/sys/ps
"make" -C /data/riotbuild/riotbase/sys/random
"make" -C /data/riotbuild/riotbase/sys/random/tinymt32
"make" -C /data/riotbuild/riotbase/sys/stdio_uart
"make" -C /data/riotbuild/riotbase/sys/timex
"make" -C /data/riotbuild/riotbase/sys/tsrb
"make" -C /data/riotbuild/riotbase/sys/xtimer
"make" -C /data/riotbuild/riotboard/esp32-olimex-evb
"make" -C /data/riotbuild/riotboard/common/esp32
"make" -C /data/riotbuild/riotcpu/esp32
"make" -C /data/riotbuild/riotcpu/esp32/freertos
"make" -C /data/riotbuild/riotcpu/esp32/periph
"make" -C /data/riotbuild/riotcpu/esp32/vendor
"make" -C /data/riotbuild/riotcpu/esp32/vendor/esp-idf
"make" -C /data/riotbuild/riotcpu/esp32/vendor/esp-idf/driver
"make" -C /data/riotbuild/riotcpu/esp32/vendor/esp-idf/esp32
"make" -C /data/riotbuild/riotcpu/esp32/vendor/esp-idf/heap
"make" -C /data/riotbuild/riotcpu/esp32/vendor/esp-idf/soc
"make" -C /data/riotbuild/riotcpu/esp32/vendor/esp-idf/spi_flash
"make" -C /data/riotbuild/riotcpu/esp32/vendor/esp-idf/wpa_supplicant
"make" -C /data/riotbuild/riotcpu/esp32/vendor/xtensa
   text    data     bss     dec     hex filename
  75281    4616    8640   88537   159d9 /data/riotbuild/riotproject/examples/hello-world/bin/esp32-olimex-evb/hello-world.elf
```

jia200x commented:
ACK on my side. @kaspar030, @cladmi ?

cladmi commented:
Question, in the `ESP32` PR you updated your a repository to have `float` printfs working. 
https://github.com/RIOT-OS/RIOT/pull/9426#issuecomment-425462057

Has it been included here ? I did not receive github updates about it, maybe it is just me, but I want to be sure it is included.

gschorcht commented:
@cladmi Yes 
```
git clone https://github.com/gschorcht/xtensa-esp32-elf.git 
```
is the Xtensa compiler suite I prepared which includes the newlib that supports ```printf``` for floats.

MrKevinWeiss commented:
So @cladmi is there anything else or can we go?

MrKevinWeiss commented:
All good.  GO!

cladmi commented:
Good for me on my concern. But I did not tested it or the referenced PR so I can't approve.

But this PR still need to receive a reviewed ack before merging.
It is not enforced here for the moment but the RIOT rule still applies.

miri64 commented:
> It is not enforced here for the moment but the RIOT rule still applies.

I activated the according branch protection for this repo's master branch.

cladmi commented:
@miri64 Thanks

jia200x commented:
& GO
